/jds/bin/diff -uprN gnome-panel-2.25.3.old/applets/notification_area/main.c gnome-panel-2.25.3/applets/notification_area/main.c
--- gnome-panel-2.25.3.old/applets/notification_area/main.c	2009-01-13 17:01:39.715172000 +0000
+++ gnome-panel-2.25.3/applets/notification_area/main.c	2009-01-13 17:02:39.481396000 +0000
@@ -251,6 +251,8 @@ applet_factory (PanelApplet *applet,
 
   panel_applet_set_background_widget (applet, GTK_WIDGET (applet));
 
+  na_tray_set_tooltip (GTK_WIDGET (applet), tray);
+
   gtk_container_add (GTK_CONTAINER (applet), GTK_WIDGET (tray));
   
 #ifndef NOTIFICATION_AREA_INPROCESS
/jds/bin/diff -uprN gnome-panel-2.25.3.old/applets/notification_area/na-tray.c gnome-panel-2.25.3/applets/notification_area/na-tray.c
--- gnome-panel-2.25.3.old/applets/notification_area/na-tray.c	2009-01-13 17:01:39.718089000 +0000
+++ gnome-panel-2.25.3/applets/notification_area/na-tray.c	2009-01-13 17:07:55.556453000 +0000
@@ -52,6 +52,11 @@ struct _NaTrayPrivate
   GtkWidget *box;
   GtkWidget *frame;
 
+  gulong query_tooltip_id;
+  gboolean tooltip_enabled;
+
+  GSList *icons;
+
   guint idle_redraw_id;
 
   GtkOrientation orientation;
@@ -117,6 +122,10 @@ tray_added (NaTrayManager *manager,
 
   gtk_box_pack_end (GTK_BOX (priv->box), icon, FALSE, FALSE, 0);
 
+  priv->icons = g_slist_append (priv->icons, icon);
+
+  priv->tooltip_enabled = FALSE;
+
   gtk_widget_show (icon);
 }
 
@@ -126,12 +135,20 @@ tray_removed (NaTrayManager *manager,
               TraysScreen   *trays_screen)
 {
   NaTray *tray;
+  NaTrayPrivate *priv;
 
   tray = g_hash_table_lookup (trays_screen->icon_table, icon);
   if (tray == NULL)
     return;
 
+  priv = tray->priv;
+
   g_assert (tray->priv->trays_screen == trays_screen);
+  priv->icons = g_slist_remove (priv->icons, icon);
+
+  if (g_slist_length (priv->icons) == 0) {
+    priv->tooltip_enabled=TRUE;
+  }
 
   g_hash_table_remove (trays_screen->icon_table, icon);
   /* this will also destroy the tip associated to this icon */
@@ -707,3 +724,36 @@ na_tray_force_redraw (NaTray *tray)
   if (priv->idle_redraw_id == 0)
     priv->idle_redraw_id = g_idle_add ((GSourceFunc) idle_redraw_cb, tray);
 }
+
+static gboolean
+na_tray_icon_query_tooltip(   GtkWidget      *applet,
+                              gint            x,
+                              gint            y,
+                              gboolean        keyboard_mode,
+                              GtkTooltip     *tooltip,
+                              gpointer        user_data)     
+{
+  NaTray        *tray = (NaTray*)user_data;
+  NaTrayPrivate *priv;
+
+  priv = tray->priv;
+
+  gtk_tooltip_set_markup (tooltip, _("Area where notification icons appear") );
+
+  return ( priv->tooltip_enabled ); 
+}
+
+void
+na_tray_set_tooltip (GtkWidget *applet, NaTray *tray)
+{
+  NaTrayPrivate *priv;
+  priv = tray->priv;
+
+  priv->tooltip_enabled=TRUE;
+
+  gtk_widget_set_has_tooltip( GTK_WIDGET(applet), TRUE );
+
+  priv->query_tooltip_id  = g_signal_connect(applet, "query-tooltip",
+                                             G_CALLBACK (na_tray_icon_query_tooltip), tray);
+}
+
